File: python-server/websocket_manager.py
"""
MCP 工具 - WebSocketManager
- 连接管理机制
- 基于消息 ID 的请求-响应模式
- 异步通信实现
- 错误处理和资源清理
"""

from typing import Dict, Optional, List
import uuid
import asyncio

from fastapi import WebSocket
from config import settings

from logger import get_logger

logger = get_logger(__name__)

class WebSocketManager:
    def __init__(self):
        self.active_connections: Dict[str, WebSocket] = {}  # {conn_id: websocket}
        self.pending_responses: Dict[str, asyncio.Future] = {}  # 存储待响应的 Future
        self.browser_names: Dict[str, str] = {}  # {conn_id: name}

    async def connect(self, websocket: WebSocket, conn_id: Optional[str] = None) -> str:
        logger.debug("正在接受 WebSocket 连接...")
        await websocket.accept()
        conn_id = conn_id or str(uuid.uuid4())  # 如果没有提供 conn_id，则生成一个
        self.active_connections[conn_id] = websocket
        logger.info(f"新连接建立，conn_id: {conn_id}")
        logger.debug(f"当前连接数: {len(self.active_connections)}")
        return conn_id

    def disconnect(self, conn_id: str):
        logger.debug(f"正在断开 WebSocket 连接..., 当前连接数: {len(self.active_connections)}")
        if conn_id in self.active_connections:
            self.active_connections.pop(conn_id)
            logger.info(f"连接断开，conn_id: {conn_id}")
        logger.debug(f"已断开 WebSocket 连接，当前连接数: {len(self.active_connections)}")

    def has_connection(self, conn_id: str) -> bool:
        """检查连接是否已存在"""
        return conn_id in self.active_connections

    async def send_message(
        self,
        message: dict,
        target_conn_id: Optional[str] = None,
        timeout: Optional[float] = None
    ) -> dict:
        """
        发送消息到指定连接（默认发送到第一个可用连接）
        - target_conn_id: 可指定目标连接的 conn_id
        """
        if not self.active_connections:
            raise ConnectionError("没有活动的 WebSocket 连接")
        logger.debug(f"正在发送消息, target_conn_id: {target_conn_id}, message: {message}")

        # 如果没有指定 conn_id，默认选择第一个连接
        websocket = (
            self.active_connections.get(target_conn_id)
            if target_conn_id
            else next(iter(self.active_connections.values()))
        )

        if not websocket:
            raise ConnectionError(f"未找到目标连接: {target_conn_id}")

        if not message.get("message_id", ""):
            # 如果消息中未包含 message_id, 则生产一个
            message_id = str(uuid.uuid4())
            message["message_id"] = message_id  # 加入唯一消息 ID
        else:
            message_id = message["message_id"]

        logger.debug(f"new message: {message}")
        future = asyncio.get_event_loop().create_future()
        self.pending_responses[message_id] = future

        try:
            await websocket.send_json(message)
            response = await asyncio.wait_for(future, timeout=timeout or settings.websocket_timeout)
            return response
        except asyncio.TimeoutError:
            raise ConnectionError("等待响应超时")
        finally:
            self.pending_responses.pop(message_id, None)

    async def handle_response(self, data: dict):
        """处理 Postman 返回的响应"""
        message_id = data.get("message_id")
        logger.debug(f"开始响应: {data}, pending_responses: {self.pending_responses}")
        if message_id in self.pending_responses:
            future = self.pending_responses[message_id]
            if not future.done():
                future.set_result(data)  # 通知 `send_message` 已收到响应

    def register_browser_name(self, conn_id: str, name: str):
        """注册浏览器名称映射"""
        if not name:
            raise ValueError("浏览器名称不能为空")
        if self.is_browser_name_active(name, exclude_conn_id=conn_id):
            raise ValueError(f"浏览器名称已连接: {name}")
        self.browser_names[conn_id] = name
        logger.info(f"浏览器已注册: conn_id={conn_id}, name={name}")

    def unregister_browser_name(self, conn_id: str):
        """注销浏览器名称映射"""
        self.browser_names.pop(conn_id, None)

    def get_browser_list(self) -> List[Dict[str, str]]:
        """获取已连接的浏览器列表"""
        return [
            {"conn_id": conn_id, "name": self.browser_names.get(conn_id, "")}
            for conn_id in self.browser_names.keys()
            if conn_id in self.active_connections
        ]

    def is_browser_name_active(self, name: str, exclude_conn_id: Optional[str] = None) -> bool:
        """检查浏览器名称是否已有活动连接"""
        for conn_id, active_name in self.browser_names.items():
            if conn_id != exclude_conn_id and active_name == name and conn_id in self.active_connections:
                return True
        return False

    def resolve_browser_conn_id(self, name: Optional[str] = None) -> str:
        """根据浏览器名称解析连接ID

        name 为空时使用默认浏览器（配置的 default_browser 或第一个连接）
        """
        if not self.active_connections:
            raise ConnectionError("没有活动的浏览器连接")

        if not name:
            # 尝试使用配置的默认浏览器
            if settings.default_browser:
                for conn_id, n in self.browser_names.items():
                    if n == settings.default_browser and conn_id in self.active_connections:
                        return conn_id
            # 返回第一个活动连接
            for conn_id in self.browser_names.keys():
                if conn_id in self.active_connections:
                    return conn_id
            raise ConnectionError("没有活动的浏览器连接")

        # 按名称查找
        for conn_id, n in self.browser_names.items():
            if n == name and conn_id in self.active_connections:
                return conn_id

        raise ConnectionError(f"未找到浏览器 '{name}'")

File: go-server/websocket_manager.go
package main

import (
	"context"
	"errors"
	"fmt"
	"sync"
	"time"

	"github.com/gorilla/websocket"
	"go.uber.org/zap"
)

// WSConn wraps a gorilla websocket with safe concurrent access.
type WSConn struct {
	conn   *websocket.Conn
	mu     sync.Mutex
	logger *zap.Logger
}

func NewWSConn(conn *websocket.Conn, logger *zap.Logger) *WSConn {
	return &WSConn{conn: conn, logger: logger}
}

func (w *WSConn) WriteJSON(v any) error {
	w.mu.Lock()
	defer w.mu.Unlock()
	return w.conn.WriteJSON(v)
}

func (w *WSConn) ReadJSON(v any) error {
	return w.conn.ReadJSON(v)
}

func (w *WSConn) Close() error {
	return w.conn.Close()
}

// BrowserInfo represents a connected browser instance.
type BrowserInfo struct {
	ConnID string `json:"conn_id"`
	Name   string `json:"name"`
}

// WebSocketManager manages active WebSocket connections and pending responses.
type WebSocketManager struct {
	mu                sync.RWMutex
	activeConnections map[string]*WSConn
	pendingResponses  map[string]chan *BrowserResponse
	browserNames      map[string]string // conn_id -> name
	logger            *zap.Logger
}

func NewWebSocketManager(logger *zap.Logger) *WebSocketManager {
	return &WebSocketManager{
		activeConnections: make(map[string]*WSConn),
		pendingResponses:  make(map[string]chan *BrowserResponse),
		browserNames:      make(map[string]string),
		logger:            logger,
	}
}

// Connect registers a new WebSocket connection.
func (wm *WebSocketManager) Connect(connID string, ws *WSConn) {
	wm.mu.Lock()
	defer wm.mu.Unlock()
	wm.activeConnections[connID] = ws
	wm.logger.Info("WebSocket connection established", zap.String("conn_id", connID))
	wm.logger.Debug("Active connection count", zap.Int("count", len(wm.activeConnections)))
}

// Disconnect removes a WebSocket connection.
func (wm *WebSocketManager) Disconnect(connID string) {
	wm.mu.Lock()
	defer wm.mu.Unlock()
	if _, ok := wm.activeConnections[connID]; ok {
		delete(wm.activeConnections, connID)
		wm.logger.Info("WebSocket connection closed", zap.String("conn_id", connID))
	}
	wm.logger.Debug("Active connection count", zap.Int("count", len(wm.activeConnections)))
}

// HasConnection checks if a connection exists.
func (wm *WebSocketManager) HasConnection(connID string) bool {
	wm.mu.RLock()
	defer wm.mu.RUnlock()
	_, ok := wm.activeConnections[connID]
	return ok
}

// RegisterBrowserName associates a human-readable name with a connection.
func (wm *WebSocketManager) RegisterBrowserName(connID, name string) error {
	if name == "" {
		return errors.New("browser name is required")
	}
	wm.mu.Lock()
	defer wm.mu.Unlock()
	for existingConnID, existingName := range wm.browserNames {
		if existingConnID != connID {
			if _, ok := wm.activeConnections[existingConnID]; ok && existingName == name {
				return fmt.Errorf("browser name already connected: %s", name)
			}
		}
	}
	wm.browserNames[connID] = name
	wm.logger.Info("Browser registered", zap.String("conn_id", connID), zap.String("name", name))
	return nil
}

// UnregisterBrowserName removes the name mapping for a connection.
func (wm *WebSocketManager) UnregisterBrowserName(connID string) {
	wm.mu.Lock()
	defer wm.mu.Unlock()
	delete(wm.browserNames, connID)
}

// GetBrowserList returns all connected browsers with their names.
func (wm *WebSocketManager) GetBrowserList() []BrowserInfo {
	wm.mu.RLock()
	defer wm.mu.RUnlock()

	list := make([]BrowserInfo, 0, len(wm.browserNames))
	for connID, name := range wm.browserNames {
		if _, ok := wm.activeConnections[connID]; ok {
			list = append(list, BrowserInfo{ConnID: connID, Name: name})
		}
	}
	return list
}

// IsBrowserNameActive checks whether a named browser is already connected.
func (wm *WebSocketManager) IsBrowserNameActive(name string) bool {
	wm.mu.RLock()
	defer wm.mu.RUnlock()
	for connID, activeName := range wm.browserNames {
		if activeName == name {
			if _, ok := wm.activeConnections[connID]; ok {
				return true
			}
		}
	}
	return false
}

// ResolveBrowserConnID finds a connection ID by browser name.
// If name is empty, returns the default connection (first active or configured default).
func (wm *WebSocketManager) ResolveBrowserConnID(name string) (string, error) {
	wm.mu.RLock()
	defer wm.mu.RUnlock()

	if len(wm.activeConnections) == 0 {
		return "", errors.New("no browser connections available")
	}

	// If no name specified, use the configured default browser name.
	if name == "" {
		defaultName := GetSettings().DefaultBrowser
		if defaultName != "" {
			for connID, n := range wm.browserNames {
				if n == defaultName {
					return connID, nil
				}
			}
			// Default configured but not found — fall through to first connection.
		}
		// Return the first active connection.
		for connID := range wm.browserNames {
			if _, ok := wm.activeConnections[connID]; ok {
				return connID, nil
			}
		}
		return "", errors.New("no browser connections available")
	}

	// Look up by name.
	for connID, n := range wm.browserNames {
		if n == name {
			if _, ok := wm.activeConnections[connID]; !ok {
				continue
			}
			return connID, nil
		}
	}
	return "", fmt.Errorf("browser '%s' not found", name)
}

// SendMessage sends a request to the target connection and waits for a response.
func (wm *WebSocketManager) SendMessage(ctx context.Context, req *BrowserRequest, targetConnID string) (*BrowserResponse, error) {
	wm.mu.RLock()
	ws, ok := wm.activeConnections[targetConnID]
	wm.mu.RUnlock()

	if !ok {
		return nil, errors.New("target WebSocket connection not found")
	}

	// Ensure message_id exists.
	if req.MessageID == "" {
		req.MessageID = generateID()
	}

	// Create a response channel.
	respCh := make(chan *BrowserResponse, 1)
	wm.mu.Lock()
	wm.pendingResponses[req.MessageID] = respCh
	wm.mu.Unlock()

	defer func() {
		wm.mu.Lock()
		delete(wm.pendingResponses, req.MessageID)
		wm.mu.Unlock()
	}()

	wm.logger.Debug("Sending message",
		zap.String("target_conn_id", targetConnID),
		zap.String("message_id", req.MessageID),
		zap.String("command", req.Command),
	)

	if err := ws.WriteJSON(req); err != nil {
		return nil, fmt.Errorf("failed to send message: %w", err)
	}

	// Wait for response with context timeout.
	timeout := time.Duration(GetSettings().WebsocketTimeout * float64(time.Second))
	if deadline, ok := ctx.Deadline(); ok {
		timeout = time.Until(deadline)
	}

	select {
	case resp := <-respCh:
		return resp, nil
	case <-time.After(timeout):
		return nil, errors.New("waiting for response timed out")
	case <-ctx.Done():
		return nil, ctx.Err()
	}
}

// HandleResponse routes a response to its pending request.
func (wm *WebSocketManager) HandleResponse(resp *BrowserResponse) {
	if resp.MessageID == "" {
		wm.logger.Warn("Response missing message_id", zap.Any("data", resp))
		return
	}

	wm.mu.Lock()
	ch, ok := wm.pendingResponses[resp.MessageID]
	wm.mu.Unlock()

	if !ok {
		wm.logger.Warn("No pending request for response", zap.String("message_id", resp.MessageID))
		return
	}

	select {
	case ch <- resp:
		wm.logger.Debug("Response delivered", zap.String("message_id", resp.MessageID))
	default:
		wm.logger.Warn("Response channel full, dropping", zap.String("message_id", resp.MessageID))
	}
}

// ReadLoop continuously reads messages from a connection and routes them.
func (wm *WebSocketManager) ReadLoop(connID string, ws *WSConn) {
	defer wm.Disconnect(connID)

	for {
		var resp BrowserResponse
		if err := ws.ReadJSON(&resp); err != nil {
			if websocket.IsUnexpectedCloseError(err, websocket.CloseGoingAway, websocket.CloseNormalClosure) {
				wm.logger.Warn("WebSocket read error", zap.String("conn_id", connID), zap.Error(err))
			}
			return
		}

		// Messages with a message_id are treated as responses.
		if resp.MessageID != "" {
			wm.HandleResponse(&resp)
			continue
		}

		// Otherwise, log and ignore (heartbeat, auth, etc.).
		wm.logger.Debug("Received non-response message", zap.String("conn_id", connID), zap.Any("data", resp))
	}
}

var idCounter uint64
var idMu sync.Mutex

func generateID() string {
	idMu.Lock()
	defer idMu.Unlock()
	idCounter++
	return fmt.Sprintf("msg-%d-%d", time.Now().UnixNano(), idCounter)
}

// CloseAll closes all active WebSocket connections.
func (wm *WebSocketManager) CloseAll() {
	wm.mu.Lock()
	defer wm.mu.Unlock()
	for connID, conn := range wm.activeConnections {
		wm.logger.Info("Closing connection during shutdown", zap.String("conn_id", connID))
		_ = conn.Close()
	}
	// Clear maps
	wm.activeConnections = make(map[string]*WSConn)
	wm.browserNames = make(map[string]string)
}
